### PR TITLE
Updates to sampling of energy and photon fluxes: bug fixes and calculate unabsorbed components

### LINKED ISSUE #799: confusion over scales parameter for sample_energy/photon_flux
The documentation for the `scales` parameter for `sample_energy_flux` (everything also holds for `sample_photon_flux`) claims that when `correlated=False` the `scales` parameter should be a 1D vector with `n` elements, and when `correlated=True` it is a 2D `n x n` matrix. This matches the underlying `sherpa.sim.sample.ParameterScaleVector/Matrix` classes that eventually get called (through a string of classes and methods). That is, these classes expect the errors (1D) or covariance matrix (2D). We do need to improve the documentation to indicate that one is an error and one a variance.

The reason for this ticket however is that `sample_energy_flux` fails if `correlated=False` and `scales` is a 1D NumPy array. That is:

```python
sherpa.astro.ui.sample_energy_flux(lo=0.5, hi=7, num=1, scales=np.asarray([0.04, 0.11, 2.4e-5]))
```

fails with the error

```
~/sherpa/sherpa-sample-flux/sherpa/astro/flux.py in sample_flux(fit, data, src, method, correlated, num, lo, hi, numcores, samples)
     90     if correlated == False and samples is not None:
     91         if numpy.ndarray == type(samples):
---> 92             samples = samples.diagonal(0)
     93
     94     samples = sampler.get_sample(fit, samples, num=num)

ValueError: diag requires an array of at least two dimensions
```

The work around is to use a Python list, since this works:

```python
sherpa.astro.ui.sample_energy_flux(lo=0.5, hi=7, num=1, scales=[0.04, 0.11, 2.4e-5])
```

The error comes because the code doesn't distinguish between 1D and 2D NumPy arrays, which it could. Or we could require a 1D array when `correlated=False` (as suggested by the docs).

### LINKED ISSUE #801: plot_energy_flux / plot_photon_flux do not use the scales parameter
`plot_energy_flux` (and the `_photon_` variant) does not follow `sample_energy_flux` and let you manually specify the error distribution to use (i.e. the `scales` parameter is not used).

This was found while inspecting the code for other changes.

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Fixes and improvements to the `energy_flux` and `photon_flux` set of commands: `sample_energy_flux`, `sample_photon_flux`, `plot_energy_flux`, `plot_photon_flux`, `get_energy_flux_hist`, and `get_photon_flux_hist`:

 - the routines can now take a model parameter to allow the flux calculation to be made on a component of the full expression. This allows you to get the flux errors for the unabsorbed model component (e.g. if the full source model is xsphabs * powlaw1d then the calculations can just be done for the powlaw1d component now). It extends #756 to the remaining `_energy_`/`_photon_` functions. As discussed below there are some consequences to this change.
 - improve support and documentation for the scales parameter (fix #799)
 - the `plot_` and `get_` functions did not support the scales parameter (fix #801)
 - support for simultaneous fits has been improved for the `sample_energy_flux`, `sample_photon_flux`, `plot_energy_flux`, `plot_photon_flux`, `get_energy_flux_hist`, and `get_photon_flux_hist` functions, as they can now specify a subset of datasets using the `otherids` parameter, as is done with other functions
 - take advantage of NumPy routines where appropriate (no user-visible changes)
 - there are now more checks on parameter values to ensure they are valid (e.g. the correct length or shape for 1D and 2D arrays, or that inputs do not contain invalid values such as None or NaN when appropriate)
 - tests have been added for existing and updated functionality
 - add docstrings to some of the routines in the `sherpa.astro.flux` module (these are not seen by users of the `sherpa.astro.ui` module).

# Details

These commits have been separated out of #754.

## Changes when sampling using a component of the full model expression

This PR now lets you fit a model like `xsphabs.gal * powlaw1d.pl` but just calculate the flux distribution for the `pl` component. 

A previous version of this PR ensured that the return array *only* contain parameter distributions for the thawed parameters in the model component used to calculate the flux. However, I have reverted this part in a later commit.

So, the return value from

    sample_energy_flux(model=pl)

will have 4 columns - that is (`flux`, `gal.nh`, `pl.gamma`, `pl.ampl`) - even though `gal.nh` isn't used to calculate the flux (as `model=pl`).

In this mode, the `scales` parameter requires the full number of parameters (so in this example 4 of them). This is to make it easy to use values calculated by `covar` or `conf`, which will include all parameters, without having to filter them to remove the "extra" parameters. It does mean that if you want to manually create the errors (e.g. to see how things change) you will need to create values for these "unused" parameters too.

## Example of calculating the flux distribution for an unabsorbed component

```
Dataset               = 1
Method                = levmar
Statistic             = chi2datavar
Initial fit statistic = 35.2038
Final fit statistic   = 35.2034 at function evaluation 9
Data points           = 42
Degrees of freedom    = 39
Probability [Q-value] = 0.643742
Reduced statistic     = 0.902651
Change in statistic   = 0.000454546
   gal.nH         0.0391826    +/- 0.036544
   pl.gamma       2.02744      +/- 0.106697
   pl.ampl        0.000196152  +/- 2.3637e-05

In [2]: plot_energy_flux(lo=0.5, hi=2, bins=20, num=2000)

In [3]: plot_energy_flux(lo=0.5, hi=2, bins=20, num=2000, overplot=True, model=pl)

In [4]: calc_energy_flux(lo=0.5, hi=2)
Out[4]: 3.8533350781298754e-13

In [5]: calc_energy_flux(lo=0.5, hi=2, model=pl)
Out[5]: 4.3570819963731163e-13

In [6]: from matplotlib import pyplot as plt

In [7]: plt.axvline(3.853e-13)
Out[7]: <matplotlib.lines.Line2D at 0x7f1dcda14f70>

In [8]: plt.axvline(4.357e-13)
Out[8]: <matplotlib.lines.Line2D at 0x7f1dcda1f8e0>
```

The blue histogram is from the first `plot_energy_flux` call and shows the result for the full source model (in this case `xsphabs * powlaw1d`) whereas the orange histogram shows the result for just the power-law component (this capability is added in this PR). The best-fit absorbed and unabsorbed fluxes are calculated with `calc_energy_flux` and then drawn on as vertical lines, "for fun". The ability to use the `model` parameter with `calc_energy_flux` is a relatively recent addition (post the CIAO 4.12.0 release).

pr-fluxes

## Example of parameter distributions

With the above fit (i.e. `xsphabs.gal * powlaw1d.pl`), we can have (s2 is new in this PR)

```
>>> s1 = sample_energy_flux(0.5, 2, num=5000)
>>> s2 = sample_energy_flux(0.5, 2, num=5000, model=pl)
```

and then view (e.g. after a `pip install corner`):

```
>>> corner.corner(s1, labels=['flux', 'nh', 'gamma', 'ampl'])
>>> corner.corner(s2, labels=['flux', 'nh', 'gamma', 'ampl'])
```

### Absorbed fluxes

pr-pardist-all

### Unabsorbed fluxes

pr-pardist-pl

### Overplotting (see below for an example of how to do this)

Orange is the unabsorbed flux. so you can see how that distribution has shifted compared to the absorbed flux, but the gamma and pl distributions are the same. Note that the tight correlation between flux and ampl for the unabsorbed case is because of the restricted energy range used here (0.5-2 keV); if I'd used 0.5-7 keV then there would be more variation (as you can see in other examples, I think).

pr-pardist-overplot

## Example with correlated parameters

This uses the covariance matrix (only thing that is new here is that `model` can be given), but note that it has used the correct sub-matrix just for the `pl` component [you can compare the widths to the previous plot and see they are similar for the parameter distributions]:

```
>>> s3 = sample_energy_flux(0.5, 7, num=5000, model=pl, correlated=True)
>>> corner.corner(s3)
```

pr-pardist-pl-correlated

## Overplotting the distributions

Here I switch to 0.5-7 keV for both uncorrelated and correlated analysis, and overplot the histogram/contours, for fun:

```
>>> x = sample_energy_flux(lo=0.5, hi=7, num=5000, model=pl, correlated=False)
>>> y = sample_energy_flux(lo=0.5, hi=7, num=5000, model=pl, correlated=True)
>>> cplot = corner.corner(x, labels=['flux', 'gamma', 'ampl'])
>>> corner.corner(y, color='orange', hist_kwargs={'ls': 'dashed'}, fig=cplot)
>>> cplot.set_size_inches(8, 8)
```
pr-overplot

You can see that the gamma and ampl parameter distributions are similar when looked at in 1D (as a histogram), but that the flux distribution is narrower when the correlations between the parameters are included.

## Example using the scales argument

One of the changes is that the scales parameter should now work (and work for all the functions). Here's an example coupled with the use of the model argument, as it provides some interesting choices. In this particular example the `covar` output is:

```
Dataset               = 1
Confidence Method     = covariance
Iterative Fit Method  = None
Fitting Method        = levmar
Statistic             = chi2datavar
covariance 1-sigma (68.2689%) bounds:
   Param            Best-Fit  Lower Bound  Upper Bound
   -----            --------  -----------  -----------
   gal.nH          0.0391826    -0.035761     0.035761
   pl.gamma          2.02744    -0.104797     0.104797
   pl.ampl       0.000196152 -2.32028e-05  2.32028e-05
```

We can send in these errors explicitly either for all three parameters, even though nh is not used in calculating the flux:

```
>>> s1 = sample_energy_flux(0.5, 2, model=pl, scales=[0.04, 0.11, 2.3e-5], num=1000)
```

The output contains columns for flux, nh, gamma, and ampl:

```
>>> s1.shape
(1000, 4)
```

The `scales` parameter could also be the covariance matrix - so a 3 by 3 matrix in this case. It has to be a matrix when `correlated=True` but can be 1D or 2D when `correlated=False` (if 2D it is correctly converted to a 1D array of sigma values)

## Example with multiple datasets

Note that the support for `otherids` provides more flexibility, and allows you to do somethings you couldn't before, but for many cases it isn't actually necessary. For example, in the following example I say

    f1 = sample_energy_flux(lo=0.5, hi=7, id=1, otherids=(3, 4), num=n)

but we could have used tthe default `id` value (of `None`), which uses **all** available datasets, ie I could have said

    f1 = sample_energy_flux(lo=0.5, hi=7, num=n)

Adding in the `otherids` argument lets you pick a subset of data to use, or use a model/dataset that is not the default one for the calculations.

The example script (here using data from sherpa's test dataset):

```
from sherpa.astro.ui import *

mdl = xsphabs.gal * powlaw1d.spl
bmdl = powlaw1d.bpl

load_pha(1, 'sherpa-test-data/sherpatest/obs1.pi')
load_pha(3, 'sherpa-test-data/sherpatest/obs3.pi')
load_pha(4, 'sherpa-test-data/sherpatest/obs4.pi')

set_source(1, mdl)
set_source(3, mdl)
set_source(4, mdl)

set_bkg_source(1, bmdl)
set_bkg_source(3, bmdl)
set_bkg_source(4, bmdl)

notice(0.5, 7)

set_stat('cstat')

gal.nh = 0.0393
gal.nh.freeze()

spl.gamma = 1.76
spl.ampl = 7.33e-7

bpl.gamma = 1.42
bpl.ampl = 1.-5e-6

fit()

n = 10000

# use all three datasets
f1 = sample_energy_flux(lo=0.5, hi=7, id=1, otherids=(3, 4), num=n)

# only use dataset 1
s1 = sample_energy_flux(lo=0.5, hi=7, id=1, num=n)
```

we can evaluate the flux for the combined dataset (called `f1` below) and just for dataset 1 (`s1`). I draw the single dataset first, so it's in black, and I haven't bothered to fix the scaling to show the range of this tataset; it's focussed on the smaller range of the "combined" dataset. A remider that the three columns in the figure below are flux, gamma, and amplitude.

```
>>> import corner
>>> fig1 = corner.corner(s1, labels=['flux', 'gamma', 'ampl', 'bgamma', 'bampl'])
>>> corner.corner(f1, color='orange', fig=fig1)
```

multi

You can see that the distributions are much narrower for the orange (ie combined) dataset. For reference the datasets have times approximately 40ks, 80ks, 40ks, so the `s1` dataset is ~1/4 the length of the combined dataset. It's not obvious to me
why corner has chosen such "interesting" scaling here for the datasets.

## Examples of error messages

For completeness (removing the extra traceback):

```
>>> sample_energy_flux(num=0)
ArgumentErr: Invalid num: 'must be a positive integer'
>>> sample_energy_flux(scales=[0.2])
ModelErr: expected 3 thawed parameters, got 1
>>> sample_energy_flux(scales=[[0.2, 0.1, 0.3], [0.1, 0.2, 0.3]])
ArgumentErr: Invalid scales: 'scales must be square when 2D'
>>> sample_energy_flux(scales=[0.2, None, 0.4])
ArgumentErr: Invalid scales: 'must not contain None values'
>>> sample_energy_flux(scales=np.asarray([0.2, 0.1, np.nan]))
ArgumentErr: Invalid scales: 'must only contain finite values'
```

```
>>> get_source()
<BinaryOpModel model instance '(xswabs.gal * powlaw1d.pl)'>
>>> sample_energy_flux(model=const1d.bgnd)
ArgumentErr: Invalid src: 'model contains term not in fit'
>>> freeze(pl)
>>> sample_energy_flux(model=pl)
FitErr: model has no thawed parameters
```

### CODE REVIEW

**Comment by DougBurke** [comment]
I need to think about the "only get back the parameters related to the model" for when the parameters are correlated case.

**Comment by DougBurke** [comment]
I've rebased this to get the codecov change to see how it looks.

**Comment by DougBurke** [comment]
@anetasie - can you link to the actual code, as it makes it a lot easier to find them (I don't know what file you are talking about).

**Comment by anetasie** [comment]
@DougBurke It is in sherpa/ui/utils.py

**Comment by DougBurke** [comment]
So

https://github.com/sherpa/sherpa/pull/803/files#diff-94a34e740f94d1a1c0b943e78b02c3f8R8651

and

https://github.com/sherpa/sherpa/pull/803/files#diff-94a34e740f94d1a1c0b943e78b02c3f8R10313

**Comment by DougBurke** [comment]
For these two changes there was no change to the code, I am just better documenting what the existing code does.

**Comment by DougBurke** [comment]
I'm about to rebase due to conflicts in sherpa/astro/ui/tests_astro_ui_plot.py. That's the problem with adding new tests; there's potential conflicts!

**Comment by hamogu** [comment]
In your first comment, section "Example of parameter distributions", I'm looking at s2. The model is "pl", so it does not include the absorption. So what's the meaning of the n_H values I see in the corner plot? They are not part of the model, so how are they drawn? I'm not asking about the mechanics, but rather the intended interpretation. Are these just dummy values that have nothing to do with the actual values of the model? Is that why they show no correlation with the flux in the "unbasorbed" case?  Yet, the values for "n_H" are not simply a featureless ellipse, but have some structure, so how is this structure determined, if it's nothing to do with the actual model?

Without going through the code, I just wonder what the output means? Would it be an alternative to return just the input model value for N_H to make clear that the distribution should not be interpreted, but still keep the dimensionality of input and output vector? Or maybe I'm not understanding what the purpose is?

**Comment by DougBurke** [comment]
The total model is used to evaluate the errors on the parameters and these are used to create the samples. So the nH values should be "truthful" (e.g. have correlations with other parameters if using correlate=False), it's just that it doesn't end up being used. It is confusing, but makes the internals easier. An earlier version of this PR removed the unused parameters like nH, but I found that code harder to maintain and reason about, and it needed more sanity checks.

There may be better ways of doing this, but I decided against the perfect being the better of the good here and wanted input on this. My comment from the commit 

> A previous version of this PR ensured that the return array only contain parameter distributions for the thawed parameters in the model component used to calculate the flux. However, I have reverted this part in a later commit.

suggests that some of the old code exists in earlier commits.

**Comment by hamogu** [comment]
I did read that comment, but it was not clear to me that the values are truthful. 
One could presumably just not plot them (something like `corner.corner(s1[1:, :], labels=['flux', 'gamma', 'ampl'])`(example only, I did not actually check the order of the parameters above).

**Comment by DougBurke** [comment]
I've rebased so that I can improve some of the tests with the reset_seed fixture. Let's hope it works on travis.

**Comment by DougBurke** [comment]
Rebased because of conflict.

**Comment by DougBurke** [comment]
@anetasie - do you still have comments left on this? If not can you finish you review. If you still have comments then please repeat them as I thought they were all addressed!

**Comment by anetasie** [comment]
@DougBurke no more question. I plan to run more tests after this PR is merged.

**Review by anetasie — APPROVED** [review]

File: sherpa/ui/utils.py
Comment (DougBurke) [reply sherpa/ui/utils.py:8697]: For all four points, all I am doing is (hopefully) better describing what the code is doing. I have not changed any code for these routines, only documentation,

File: sherpa/ui/utils.py
Comment (DougBurke) [reply sherpa/ui/utils.py:10360]: For all four points, all I am doing is (hopefully) better describing what the code is doing. I have not changed any code for these routines, only documentation,

File: sherpa/ui/utils.py
Comment (DougBurke) [reply sherpa/ui/utils.py:8643]: For all four points, all I am doing is (hopefully) better describing what the code is doing. I have not changed any code for these routines, only documentation,

File: sherpa/ui/utils.py
Comment (DougBurke) [reply sherpa/ui/utils.py:8698]: For all four points, all I am doing is (hopefully) better describing what the code is doing. I have not changed any code for these routines, only documentation,

**Review by dtnguyen2 — COMMENTED** [review]
There's a type on line 385 of the file _sherpa/astro/flux.py_:
 > numcores : int or None, optonal

On line 395 of the file _sherpa/astro/flux.py_:
>       2D array, and contain the covariance matrix for the free
I can't remember anymore but doesn't the covariance matrix have to be positive (semi-)definite?

**Comment by DougBurke** [comment]
@dtnguyen2 - I honestly have no idea about the requirements of the covariance matrix. It would be good to document requirements like that, but I don't want to hold this back (since the previous version of the docs never really mentioned it either, I believe) ;-)

**Comment by DougBurke** [comment]
@dtnguyen2 - tracing through the code, one code path hits this check

        # Investigate spectral decomposition to avoid requirement that the cov be
        # semi-positive definite.  Nevermind, NumPy already uses SVD to generate
        # deviates from a multivariate normal.  An alternative is to use Cholesky
        # decomposition, but it assumes that the matrix is semi-positive
        # definite.
        if numpy.min(numpy.linalg.eigvalsh(cov)) <= 0:
            raise TypeError("The covariance matrix is not positive definite")

so I can add a note about this (even if it isn't always needed)

**Comment by DougBurke** [comment]
@dtnguyen2 - I've made the doc changes.

**Comment by anetasie** [comment]
> @dtnguyen2 - I honestly have no idea about the requirements of the covariance matrix. It would be good to document requirements like that, but I don't want to hold this back (since the previous version of the docs never really mentioned it either, I believe) ;-)

@DougBurke the definition of the covariance matrix is that it is symmetric and positive semi-definite, so if this condition is not set then the matrix is not a covariance matrix. 
[https://en.wikipedia.org/wiki/Covariance_matrix]

**Comment by DougBurke** [comment]
@anetasie - the trouble is that I believe we can return a "covariance matrix" from our error analysis thanks to numerical issues, so it's probably worthwhile just mentioning it as a constraint.

**Comment by anetasie** [comment]
Sure.

**Review by dtnguyen2 — APPROVED** [review]

File: sherpa/astro/flux.py
Comment (wmclaugh) [inline sherpa/astro/flux.py:108]: sapmles -> samples

Comment (DougBurke) [reply sherpa/astro/flux.py]: Thanks - fixed.

**Comment by dtnguyen2** [comment]
sherpa's ```covariance``` implementation is very troublesome.  One of these days, perhaps we should look into if it can be improved or not

**Pull request merged**
